src/core/thread/mesh_forwarder.cpp

```diff
@@ -1096,6 +1096,17 @@ void MeshForwarder::HandleReceivedFrame(Mac::RxFrame &aFrame)
             LogFrame("Received empty payload frame", aFrame, OT_ERROR_NONE);
         }
 
+#if OPENTHREAD_THREAD_VERSION >= OPENTHREAD_THREAD_VERSION_1_2 && OPENTHREAD_FTD
+        // From Thread 1.2, MAC Data Frame can also act as keep-alive message if child supports
+        {
+            Child *child = Get<ChildTable>().FindChild(macSource, ChildTable::kInStateAnyExceptInvalid);
+
+            if (child != NULL && child->IsEnhancedKeepAliveSupported())
+            {
+                child->SetLastHeard(TimerMilli::GetNow());
+            }
```

Comment:
The MAC Data Frame should only be used as a keep alive if the child is rx-off-when-idle, rx-on children are required to use the Child Update Request keep-alive.

Reply:
I am curios about why spec kept non-sleepy same as before.

If we receive a secure data frame from a child (sleepy or not), would it not indicate that child is still connected and ok (which I think was intention behind the keep-alive msg)?

Reply:
It is really down to a zero-compromises approach. Going back to Thread 1.1, the Child Update Request proves a fully functional bi-directional link including security, because the ChildUpdateRequest made it up to the parent, was processed, and the ChildUpdateResponse made it back to the child. The MAC Data request never did that, as a parent will Ack a MAC Data Request even if it can't process the security or doesn't consider it a child. This is why we have the sentence:

> Rx-off-when-idle Children SHOULD also periodically (at time intervals indicated by the application) resynchronize by sending a Child Update Request to their Parent to avoid cases when the MAC Data Request is being acknowledged erroneously by a receiver that is not a valid Parent.

In Thread 1.2, we introduced the Enhanced Frame Pending feature, which allows a child to accomplish the same thing as a MAC Data Request command (more or less) with a MAC Data frame - be aware of when the parent has data. Therefore using a MAC Data frame here is strictly an improvement vs Thread 1.1, whereas making it possible for Rx-On children would be a trade off.

Reply:
Thanks @CiaranWoodward for explanation on this.